plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -340,3 +340,36 @@ instance KnownBuiltinType term String        => KnownType term String
 instance KnownBuiltinType term Char          => KnownType term Char
 instance KnownBuiltinType term ()            => KnownType term ()
 instance KnownBuiltinType term Bool          => KnownType term Bool
+
+{- Note [Int as Integer]
+We represent 'Int' as 'Integer' in PLC and check that an 'Integer' fits into 'Int' when
+unlifting constants (and fail if it doesn't).
```

Comment:
This seems wrong to me. I thought that rather:
- Our `Integer` builtin type *really is* `Integer`: it's arbitrary-precision. We don't want to deal with overflow issues most of the time!
- *Some* of our builtins don't handle `Integer`, they really need `Int`. Mostly the bytestring ones. So we should do a restricting conversion `Integer -> Int` at the point where we take the argument, and the builtin should fail (as builtins normally fail) if that conversion doesn't work.

Reply:
In other words, the previous behaviour seemed right, we just needed to handle the case where the restriction doesn't work in the semantics of the builtin function.

Reply:
Err, stupid docs, sorry. This sentence should be

> We represent 'Int' as 'Integer' in PLC and check that an 'Integer' fits into 'Int' when
unlifting constants of type 'Int' (and fail if it doesn't).

(added the extra "of type 'Int').

If you look at the code, what happens there is exactly what you described:

> So we should do a restricting conversion `Integer -> Int` at the point where we take the argument, and the builtin should fail (as builtins normally fail) if that conversion doesn't work.

Reply:
`Integer` is still a separate type and lifting from / unlifting to `Integer` works as before. I only added an instance for `Int` that does the checking.

Reply:
But we're still doing the failure "outside" the builtin semantics, right? Which is why you have this whole problem with the errors. What I'm suggesting is we write

```
specialTake :: Integer -> ByteString -> ByteString
specialTake i bs | tooSmall i = ... -- whatever you do to indicate that builtins fail, I can't remember
specialTake i bs = take (fromIntegral i) bs
```

Reply:
So then the builtin "really" takes an `Integer`, so we have no need to pretend that we can unlift `Int`s.

Reply:
Yeah, we can do that, but I kinda like to pretend we can unlift `Int`s as that's what we really do in the end. Ok, I agree broken semantics is not worth it, I'll try implementing the `AsEvaluationFailure` thing, it should improve other parts as well, so maybe that'll convince you.